lib/jsonrpc_client.ml

```diff
@@ -17,49 +17,95 @@
+				| 0 -> Buffer.contents buf (* EOF *)
+				| n ->
+					if n > max_bytes
+					then
+					begin
+						debug "exceeding maximum read limit %d, clear buffer" !json_rpc_max_len;
+						Buffer.clear buf;
+						raise Read_error
+					end
+					else
+					begin
+						Buffer.add_subbytes buf bytes 0 n;
+						inner remain_time (max_bytes - n)
+					end
+				| exception Unix.Unix_error(err,_,_) when err = Unix.EAGAIN || err = Unix.EWOULDBLOCK ->
+					inner remain_time max_bytes
+		end
+		else inner remain_time max_bytes
 	in
-	get ();
-	Buffer.contents buf
-
-let receive fin =
-	let obj = input_json_object fin in
-	debug "Response: %s" obj;
-	Jsonrpc.response_of_string obj
+	inner timeout !json_rpc_max_len
 
-let with_connection sockaddr f =
-	let fin, fout = Unix.open_connection sockaddr in
-	debug "Connected.";
-	let result = f fin fout in
-	Unix.shutdown_connection fin;
-	close_in fin;
-	debug "Shut down.";
-	result
+(* Write as many bytes to a file descriptor as possible from data before a given clock time. *)
+(* Raises Timeout exception if the number of bytes written is less than the specified length. *)
+(* Writes into the file descriptor at the current cursor position. *)
+let timeout_write filedesc total_length data response_time =
+	let write_start = Mtime_clock.counter () in
+	let get_total_used_time () = Mtime.Span.to_uint64_ns (Mtime_clock.count write_start) in
+	let rec inner_write filedesc offset length remain_time =
+		let (_, ready_to_write, _) = Unix.select [] [filedesc] [] (to_s remain_time) in 
+		let used_time = get_total_used_time () in
+		let new_remain_time = Int64.sub response_time used_time in
+		if new_remain_time < 0L then
+		begin
+			debug "Timeout to write %d at offset %d" total_length offset;
+			raise Timeout
+		end;
+		if List.mem filedesc ready_to_write then 
+		begin
+			let bytes_written = 
+				(try Unix.single_write filedesc data offset length with 
+				| Unix.Unix_error(Unix.EAGAIN,_,_)
+				| Unix.Unix_error(Unix.EWOULDBLOCK,_,_) -> 0)
+			in
+			let new_offset = offset + bytes_written in
+			let new_length = length - bytes_written in
+			if new_length = 0 then ()
+			else inner_write filedesc new_offset new_length new_remain_time
+		end
+		else inner_write filedesc offset length new_remain_time
+	in
+	inner_write filedesc 0 total_length response_time
 
 let with_rpc ?(version=Jsonrpc.V2) ~path ~call () =
-	let sockaddr = Unix.ADDR_UNIX path in
-	with_connection sockaddr (fun fin fout ->
-		let req = Jsonrpc.string_of_call ~version call in
-		debug "Request: %s" req;
-		output_string fout req;
-		flush fout;
-		receive fin
-	)
-
+	let uri = Uri.of_string (Printf.sprintf "file://%s" path) in
+	Open_uri.with_open_uri uri (fun s ->
+		Unix.set_nonblock s;
+		let req = Bytes.of_string (Jsonrpc.string_of_call ~version call) in
+		timeout_write s (Bytes.length req) req (Int64.of_string !json_rpc_write_timeout);
+		let res = timeout_read s (Int64.of_string !json_rpc_read_timeout) in
+		debug "Response: %s" res;
```

Comment:
If print the response, it is better to print the request at same time.

Reply:
Are we sure we want to print either of them? How often do we call this function?